lib/elixir/pages/anti-patterns/design-anti-patterns.md

```diff
@@ -411,6 +407,94 @@ end
 
+## Scattered side-effects
+
+#### Problem
+
+Elixir code tends to be explicit: it receives all inputs as arguments and returns all outputs. Furthermore, a given piece of code is clearer if, when called with the same inputs, it always returns the same output, without performing any other change in the system.
+
+For example, `String.split/2` fits this definition. It receives three arguments and, if the arguments do not change, the result is always the same:
+
+```elixir
+iex> String.split("hello world", " ")
+["hello", "world"]
+iex> String.split("hello world", " ")
+["hello", "world"]
+```
+
+On the other hand, `System.unique_integer/0` returns a different value, every time we call it:
+
+```elixir
+iex> System.unique_integer()
+-576460752303423478
+iex> System.unique_integer()
+-576460752303423446
+```
+
+We say that the `System.unique_integer/0` is *impure*, as it contains side-effects. When you invoke it, it changes the state of the program or the environment in a way that is not reflected on the output. On the other hand, we say `String.split/2` is *pure*. As a general rule, *pure* functions are easier to debug, test, and maintain then *impure* ones.
+
+Side-effects are essential when writing production software. Reading a file, logging a message, sending messages between processes, etc. are all side-effects. Writing code with side-effects is not an anti-pattern. However, you must avoid scattering side-effects throughout your functions.
+
+#### Example
+
+Consider a functionality that must parse JSON files from disk, convert all keys to lowercase, and write the updated version to the same file. An initial implementation would look like this:
+
+```elixir
+defmodule Recode do
+  require Logger
+
+  def recode!(file) do
+    data = read_and_decode!(file)
+    data = transform_data(data)
+    encode_and_write!(file, data)
+  end
+
+  defp transform_data(data) do
+    Map.new(data, fn {k, v} -> {String.downcase(k), v} end)
+  end
+
+  defp read_and_decode!(file) do
+    file |> File.read!() |> Jason.decode!()
+  end
+
+  defp encode_and_write!(file, data) do
+    File.write!(file, Jason.encode!(data))
+  end
+end
+```
+
+The code above has side-effects, such as reading and writing files, scattered around different functions. This makes the code harder to understand and maintain. For instance, as the `transform_data/1` function grows in complexity, the only way to test it is by writing and reading data from disk, which makes tests more complex and expensive then necessary.
+
+#### Refactoring
+
+To refactor this anti-pattern, we want break our code into smaller functions without side-effects, and rely on side-effects only when connecting all logic together. To do so, we define a new `recode_contents!` function, which works directly on the contents of a file, instead of file themselves.
```

Comment:
```suggestion
To refactor this anti-pattern, we want break our code into smaller functions without side-effects, and rely on side-effects only when connecting all logic together. To do so, we define a new `recode_contents!/1` function, which works directly on the contents of a file, instead of file themselves.
```